File: 프로그래머스/2/87946. 피로도/피로도.py
import itertools

def solution(k, dungeons):
    
    permutation_list = list(itertools.permutations(dungeons))
    
    answer = []   
    
    for p in permutation_list: 	
        total = 0  
        k_copy = k  
        for i, j in p:
            if i > k_copy:
                break
            else:
                k_copy = k_copy-j
                total += 1 
                
        answer.append(total)

    return max(answer)

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = 0;
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        
        visited = new boolean[dungeons.length];
        
        dfs(0, k, dungeons);
        return answer;
    }
    
    private void dfs(int count, int fatigue, int[][] dungeons) {
        for(int i = 0; i < dungeons.length; i++) {
            if(visited[i] || dungeons[i][0] > fatigue) {
                continue;
            }
            else {
                visited[i] = true;
                dfs(count+1, fatigue - dungeons[i][1], dungeons);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, count);
    }
}